From 417259a860d4f3d2ccc8190788121032cc7bb2e3 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Fri, 28 Jun 2024 08:26:48 +0300
Subject: [PATCH] [FSW-420] fs1: add spi manifest validation

with this patch, added support to validate key,
signature and firmware version check for spi
manifest.

Change-Id: Ib09037eccf5ca1e250af7ad3e98b4c21f29de004
---

diff --git a/app/boot_rom.c b/app/boot_rom.c
index 7f0d945..81e57b3 100644
--- a/app/boot_rom.c
+++ b/app/boot_rom.c
@@ -6,6 +6,7 @@
 
 #include <boot_rom.h>
 #include <security.h>
+#include <recovery.h>
 
 struct boot_rom_global_s __attribute__((section (".rom_persistent_section")))boot_rom_global = {
 	.config_fuse = 0,
@@ -34,6 +35,9 @@
 
 	boot_rom_check_n_enter_recovery();
 
+	if (boot_rom_verify_image(HL_FW_BOOT_PTR_SPI))
+		execute_recovery();
+
 	boot_rom_init_boot_info();
 
 	boot_rom_bringup_ppboot();
diff --git a/common/common.c b/common/common.c
index 5226df4..07fbb7a 100644
--- a/common/common.c
+++ b/common/common.c
@@ -16,6 +16,8 @@
 #include "dev_uart.h"
 #include <boot_desc.h>
 #include <dma.h>
+#include <secure_boot.h>
+#include <dev_security.h>
 
 void boot_rom_uart_init(void)
 {
@@ -273,3 +275,56 @@
 skip_bringup:
 	execute_recovery();
 }
+
+int boot_rom_verify_image(enum hl_fw_boot_ptr_imgs img_type)
+{
+	enum boot_rom_progress boot_desc_verif_base = PROGRESS_SPI_BOOT_DESC_VERIFY_START;
+	struct hl_fw_boot_ptrs boot_ptr;
+	uint32_t boot_ptr_addr = 0;
+	int ret;
+	uint32_t img_plus_key_len;
+
+	if (!security_efuse_image_validation_en())
+		return BOOT_ROM_SUCCESS;
+
+	if ((img_type == HL_FW_BOOT_PTR_SPI) && (boot_rom_global.validation_pass))
+		return BOOT_ROM_SUCCESS;
+
+	if (img_type == HL_FW_BOOT_PTR_RECOVERY)
+		boot_desc_verif_base = PROGRESS_RECOVERY_IMG_BOOT_DESC_VERIFY_START;
+
+	boot_ptr_addr = boot_desc_param_get(img_type, &boot_ptr);
+
+	if (!boot_ptr_addr)
+		return BOOT_ROM_FAIL;
+
+	boot_rom_status_update_bringup(boot_desc_verif_base, PROGRESS_VERIFY_START);
+
+	/*
+	 * TODO: verify it once new SPI DMA changes finalised and working.
+	 * Now we have PPBOOT is less than 64KB, so this changes expected to work.
+	*/
+	img_plus_key_len = boot_ptr.boot_ptrs_img_sz - boot_ptr.sig_sz - CRC32_SIZE;
+
+	ret = secure_boot_verify_image_signature((uint8_t *)boot_ptr_addr, img_plus_key_len,
+			(uint8_t *)(boot_ptr_addr + img_plus_key_len - boot_ptr.pub_key_sz),
+			boot_ptr.pub_key_sz, (uint8_t *)(boot_ptr_addr + img_plus_key_len), boot_ptr.sig_sz,
+			boot_ptr.sec_ver, boot_ptr.key_idx);
+
+	if (ret == SECUREBOOT_VERIFY_SUCCESS) {
+		boot_rom_status_update_bringup(boot_desc_verif_base, PROGRESS_VERIFY_SUCCESS);
+	} else {
+		if (ret == SECUREBOOT_VERIFY_VER_FAILED)
+			boot_rom_status_update_bringup(boot_desc_verif_base, PROGRESS_VERIFY_VERSION_FAIL);
+		else if (ret == SECUREBOOT_VERIFY_KEY_REVOKED)
+			boot_rom_status_update_bringup(boot_desc_verif_base, PROGRESS_VERIFY_KEY_REVOKED);
+		else if (ret == SECUREBOOT_VERIFY_KEY_FAILED)
+			boot_rom_status_update_bringup(boot_desc_verif_base, PROGRESS_VERIFY_KEY_HASH_FAIL);
+		else
+			boot_rom_status_update_bringup(boot_desc_verif_base, PROGRESS_VERIFY_AUTH_FAIL);
+
+		return BOOT_ROM_FAIL;
+	}
+
+	return BOOT_ROM_SUCCESS;
+}
diff --git a/include/boot_rom.h b/include/boot_rom.h
index 966e22c..b7051a9 100644
--- a/include/boot_rom.h
+++ b/include/boot_rom.h
@@ -119,5 +119,6 @@
 void boot_rom_check_n_enter_recovery(void);
 void boot_rom_bringup_ppboot(void);
 void boot_rom_init_boot_info(void);
+int boot_rom_verify_image(enum hl_fw_boot_ptr_imgs img_type);
 
 #endif /* _BOOT_ROM_H_ */
